Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightedGraphSchemeProtocol.swift

```diff
@@ -21,25 +23,84 @@ extension WeightedGraphSchemeProtocol {
         H: WeightedGraphSchemeProtocol,
         H.Weight == Weight
     {
-        return H.init { self.weight(Edge(f($0.a),f($0.b))) }
+        return H { self.weight(from: f($0.a), to: f($0.b)) }
     }
-}
 
-extension WeightedGraphSchemeProtocol {
     @inlinable
     func unweighted <H> () -> H where
         H: UnweightedGraphSchemeProtocol,
         H.Edge == Edge
     {
-        return H.init { self.weight($0) != nil }
+        return H { self.weight($0) != nil }
+    }
+}
+
+extension WeightedGraphSchemeProtocol where Self: UndirectedGraphSchemeProtocol, Weight: Numeric {
+    
+    static func * (lhs: Self, rhs: Self) -> Self {
+        return Self { edge in
+            guard let lweight = lhs.weight(edge), let rweight = rhs.weight(edge) else { return nil }
+            return lweight * rweight
+        }
+    }
+    
+    static func + (lhs: Self, rhs: Self) -> Self {
+        return Self { edge in
+            if let lweight = lhs.weight(edge) {
+                if let rweight = rhs.weight(edge) { return lweight + rweight }
+                return lweight
+            } else {
+                guard let rweight = rhs.weight(edge) else { return nil }
+                return rweight
+            }
+        }
     }
 }
 
-extension WeightedGraphProtocol {
-    func weightScheme <G> (_ f: @escaping (G.Node) -> Node) -> WeightCarrying<G> where
-        G: WeightedGraphProtocol,
-        G.Weight == Weight
+extension WeightedGraphSchemeProtocol where Self: DirectedGraphSchemeProtocol, Weight: Numeric {
+    
+    static func * <Scheme> (lhs: Self, rhs: Scheme) -> Self where
+    Scheme: WeightedGraphSchemeProtocol,
+    Scheme.Node == Node,
+    Scheme.Weight == Weight
     {
-        return WeightCarrying.build(from: self).pullback(f)
+        return Self { edge in
+            guard
+                let lweight = lhs.weight(edge),
+                let rweight = rhs.weight(from: edge.a, to: edge.b) else { return nil }
+            return lweight * rweight
+        }
+    }
+    
+    static func * <Scheme> (lhs: Self, rhs: Scheme) -> Self where
+        Scheme: UnweightedGraphSchemeProtocol,
+        Scheme.Node == Node
+    {
+        return Self { edge in
```

Comment:
I feel that a couple branches of logic could be swept away here. It would require adding this (which has been discussed for the stdlib [here](https://forums.swift.org/t/add-filter-and-ifsome-then-to-optional/8143/29)):

```Swift
extension Optional {
    func filter (_ predicate: (Wrapped) -> Bool) -> Optional {
        guard case let .some(wrapped) = self else { return nil }
        return predicate(wrapped) ? wrapped : nil
    }
}
```

We could then rewrite this as:

```Swift
return lhs.weight(edge).filter { _ in rhs.contains(from: edge.a, to: edge.b) }
```

Without adding functionality, we could pivot this like:

```Swift
return rhs.contains(from: edge.a, to: edge.b) ? lhs.weight(edge) : nil
```

Reply:
whoa!